homeassistant/components/guardian/switch.py

```diff
@@ -12,10 +20,49 @@
 ATTR_INST_CURRENT_DDT = "instantaneous_current_ddt"
 ATTR_TRAVEL_COUNT = "travel_count"
 
+SERVICE_DISABLE_AP = "disable_ap"
+SERVICE_ENABLE_AP = "enable_ap"
+SERVICE_REBOOT = "reboot"
+SERVICE_RESET_VALVE_DIAGNOSTICS = "reset_valve_diagnostics"
+SERVICE_UPGRADE_FIRMWARE = "upgrade_firmware"
+
+SERVICE_UPGRADE_FIRMWARE_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_URL, default=DEFAULT_FIRMWARE_UPGRADE_URL): cv.url,
+        vol.Optional(CONF_PORT, default=DEFAULT_FIRMWARE_UPGRADE_PORT): cv.port,
+        vol.Optional(
+            CONF_FILENAME, default=DEFAULT_FIRMWARE_UPGRADE_FILENAME
+        ): cv.string,
+    }
+)
+
 
 async def async_setup_entry(hass, entry, async_add_entities):
     """Set up Guardian switches based on a config entry."""
     guardian = hass.data[DOMAIN][DATA_CLIENT][entry.entry_id]
+
+    platform = entity_platform.current_platform.get()
+
+    for service_name, method in [
+        (SERVICE_DISABLE_AP, "async_disable_ap"),
+        (SERVICE_ENABLE_AP, "async_enable_ap"),
+        (SERVICE_REBOOT, "async_reboot"),
+        (SERVICE_RESET_VALVE_DIAGNOSTICS, "async_reset_valve_diagnostics"),
+    ]:
+        platform.async_register_entity_service(service_name, None, method)
+
+    async def async_handle_firmware_update(entity, call):
```

Comment:
Why does this needs it's own function? The function is called with `**call.data`, so url, port and filename just map.

Reply:
I'm not clear on what ^^^ means. Does 84ea2cf cover it?

Reply:
Did you try running it? 😝